# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/JSScript.h
# Commit: 554b7637fe60
# Full Hash: 554b7637fe60e8a57e5180b5daaa1441f1bea31b
# Author: Denis Palmeiro <dpalmeiro@mozilla.com>
# Date: 2020-05-13 03:44:05
# Regressor Bug: 1501608
# File Overlap Count: 1
# Description:
#   Bug 1501608 - Remove the ELEMENT_SLOT in the ScriptSourceObject and instead use a callback function to return the script element based on the value of the privateValue in the SSO. r=jonco,smaug
#   
#   We can reduce the size of the SSO by removing the element slot entirely, and instead retrieve the element through a callback function.  The callback will take in the value in the private slot of the SSO, which is either a LoadedScript* (from the browser) or a JSObject* (from the shell).  In addition, this removes the requirement of having a script dom element ready when parsing a JS script which can open up new opportunities for performance.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D70417
# ==============================================================================

diff -r 3470d926c02a -r 554b7637fe60 js/src/vm/JSScript.h
--- a/js/src/vm/JSScript.h	Tue May 12 09:40:48 2020 +0000
+++ b/js/src/vm/JSScript.h	Tue May 12 19:28:15 2020 +0000
@@ -1222,7 +1222,6 @@
 
   static bool initElementProperties(JSContext* cx,
                                     HandleScriptSourceObject source,
-                                    HandleObject element,
                                     HandleString elementAttrName);
 
   bool hasSource() const { return !getReservedSlot(SOURCE_SLOT).isUndefined(); }
@@ -1230,9 +1229,8 @@
     return static_cast<ScriptSource*>(getReservedSlot(SOURCE_SLOT).toPrivate());
   }
 
-  JSObject* unwrappedElement() const {
-    return unwrappedCanonical()->getReservedSlot(ELEMENT_SLOT).toObjectOrNull();
-  }
+  JSObject* unwrappedElement(JSContext* cx) const;
+
   const Value& unwrappedElementAttributeName() const {
     const Value& v =
         unwrappedCanonical()->getReservedSlot(ELEMENT_PROPERTY_SLOT);
@@ -1260,7 +1258,6 @@
   enum {
     SOURCE_SLOT = 0,
     CANONICAL_SLOT,
-    ELEMENT_SLOT,
     ELEMENT_PROPERTY_SLOT,
     INTRODUCTION_SCRIPT_SLOT,
     PRIVATE_SLOT,